File: itx-rs/itx-impl/src/queue/rabbitmq/factory.rs
use std::sync::Arc;

use itx_contract::queue::MessageQueue;
use itx_contract::queue::factory::MessageQueueFactory;
use lapin::{Connection, ConnectionProperties};

use crate::queue::rabbitmq::RabbitMessageQueue;

#[derive(Clone, serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct RabbitMessageQueueFactoryProps {
    pub host: String,
    pub port: u16,
    pub user: String,
    pub password: String,
    pub max_concurrency: u32,
    pub control_standard_queue: String,
    pub control_premium_queue: String,
    pub compute_standard_queue: String,
    pub compute_premium_queue: String,
}

pub struct RabbitMessageQueueFactory {
    conn: Arc<Connection>,
    props: RabbitMessageQueueFactoryProps,
}

impl RabbitMessageQueueFactory {
    pub async fn new(props: RabbitMessageQueueFactoryProps) -> Result<Self, lapin::Error> {
        let url = format!(
            "amqp://{}:{}@{}:{}/%2F",
            props.user, props.password, props.host, props.port
        );
        let conn = Connection::connect(&url, ConnectionProperties::default()).await?;

        Ok(Self {
            conn: Arc::new(conn),
            props,
        })
    }
}

impl MessageQueueFactory for RabbitMessageQueueFactory {
    fn create_control_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.props.control_standard_queue.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_control_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.props.control_premium_queue.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_compute_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.props.compute_standard_queue.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_compute_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(RabbitMessageQueue::new(
            self.conn.clone(),
            self.props.compute_premium_queue.clone(),
            self.props.max_concurrency,
        ))
    }
}

File: itx-go/itx-impl/queue/rabbitmq/factory.go
package rabbitmq

import (
	"fmt"

	"github.com/chehsunliu/itx/itx-go/itx-contract/queue"
	amqp "github.com/rabbitmq/amqp091-go"
)

type FactoryProps struct {
	Host                 string `yaml:"host"`
	Port                 int    `yaml:"port"`
	User                 string `yaml:"user"`
	Password             string `yaml:"password"`
	MaxConcurrency       int64  `yaml:"max-concurrency"`
	ControlStandardQueue string `yaml:"control-standard-queue"`
	ControlPremiumQueue  string `yaml:"control-premium-queue"`
	ComputeStandardQueue string `yaml:"compute-standard-queue"`
	ComputePremiumQueue  string `yaml:"compute-premium-queue"`
}

type MessageQueueFactory struct {
	conn  *amqp.Connection
	props FactoryProps
}

func New(props FactoryProps) (*MessageQueueFactory, error) {
	url := fmt.Sprintf("amqp://%s:%s@%s:%d/%%2F", props.User, props.Password, props.Host, props.Port)
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, err
	}
	return &MessageQueueFactory{conn: conn, props: props}, nil
}

func (f *MessageQueueFactory) CreateControlStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.props.ControlStandardQueue, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateControlPremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.props.ControlPremiumQueue, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateComputeStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.props.ComputeStandardQueue, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateComputePremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.conn, f.props.ComputePremiumQueue, f.props.MaxConcurrency)
}
